crate/server/src/core/operations/validate.rs

```diff
@@ -0,0 +1,586 @@
+    root: &[u8],
+    certificates: &Vec<Vec<u8>>,
+    hm_certificates: &mut HashMap<Vec<u8>, u8>,
+    _count: u8,
+) -> KResult<(ValidityIndicator, u8)> {
+    let root_x509 = X509::from_der(root)?;
+    let son_issuer_id = root_x509
+        .subject_key_id()
+        .expect("mandatory information")
+        .as_slice()
+        .to_vec();
+    let son_idx = hm_certificates.get(&son_issuer_id);
+    println!("looking for son certificate");
+    // If there is no certificate son in the vector, the iteration on the indexed
+    // structure ends returning valid.
+    if son_idx.is_none() {
+        println!("CHAIN STRUCTURE IS VALID IN: validate_chain_structure. no son_idx");
+        return KResult::Ok((ValidityIndicator::Valid, _count + 1))
+    };
+
+    // safe unwrap by construction
+    let son_cert = certificates
+        .get(*son_idx.expect("By construction, the certificate must exist in the vector") as usize);
+    if son_cert.is_none() {
+        println!("CHAIN STRUCTURE IS INVALID IN: validate_chain_structure. no son_cert");
+        return KResult::Ok((ValidityIndicator::Invalid, _count + 1))
+    };
+    let son_cert = son_cert.expect("The son certificate must exist");
+    println!("Parsing X509 certificate");
+    let son_x509 = X509::from_der(son_cert)?;
+    let root_pkey = root_x509.public_key()?;
+    let validity = son_x509.verify(&root_pkey)?;
+    let (res, count) =
+        validate_chain_structure(son_cert, certificates, hm_certificates, _count + 1)?;
+    if ValidityIndicator::from_bool(validity).and(res) == ValidityIndicator::Valid {
+        println!("CHAIN STRUCTURE IS VALID IN: validate_chain_structure");
+        KResult::Ok((ValidityIndicator::Valid, count))
+    } else {
+        println!("CHAIN STRUCTURE IS INVALID IN: validate_chain_structure");
+        KResult::Ok((ValidityIndicator::Invalid, count))
+    }
+}
+
+// If fetching a certificate fails, the method reports the first error happening.
+// Otherwise it returns a vector of certificates.
+async fn certificates_by_uid(
+    unique_identifiers: Vec<UniqueIdentifier>,
+    kms: &KMS,
+    user: &str,
+    params: Option<&ExtraDatabaseParams>,
+) -> KResult<Vec<Vec<u8>>> {
+    let res = join_all(unique_identifiers.iter().map(|unique_identifier| async {
+        let unique_identifier = unique_identifier
+            .as_str()
+            .expect("as_str returned None in certificates_by_uid");
+        certificate_by_uid(unique_identifier, kms, user, params).await
+    }))
+    .await;
+
+    // filtering errors
+    let is_err = res.iter().all(|x| x.is_err());
+
+    // checking if there are any errors
+    if !is_err {
+        KResult::Ok(
+            res.iter()
+                .map(|x| x.clone().expect("safe unwrap"))
+                .collect(),
+        )
+    } else {
+        KResult::Err(KmsError::from(KmipError::ObjectNotFound(
+            "There is a Kms Certificate UID that cannot be retrieved".to_string(),
+        )))
+    }
```

Comment:
I think there is a bug here: your variable `is_err` is `true` if *no* UID could be fetched. Indeed, since `x.is_err()` returns `false` when it is `Ok`, folding using `all` means the presence of at least one `Ok` element makes `is_err` `false`.

Reply:
oh yep!